### PR TITLE
improve internal maskings APIs

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/1358

Improve internal maskings APIs:

- previously, the mask(...) functions all returned a VPackValue to the caller. The VPackValue could contain a string, which could point to memory allocated somewhere in the returned function. To avoid pointing into invalid memory, the API also provided an std::string by-reference parameter, which the mask(...) function could fill with a value, and then return a VPackValue pointing into this string. While this worked, it was awkward and error-prone. Now, the mask(...) functions can all directly add data to the output, without having to worry about the lifetime of objects.
- the maskings API for strings was based on std::string, which required constructing temporary std::strings from velocypack string values. This is now avoided by using std::string_view in the maskings API for strings.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: -
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -

#### Related Information

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1358
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: lib/Maskings/Collection.cpp [hunk 1]
```diff
@@ -23,23 +23,26 @@
...
         "expecting an object for collection definition");
   }
 
-  std::string type = "";
+  std::string type;
   std::vector<AttributeMasking> attributes;
 
-  for (auto const& entry : VPackObjectIterator(def, false)) {
+  for (auto entry : VPackObjectIterator(def, false)) {
     std::string key = entry.key.copyString();
```
Comment (MBkkt) [inline lib/Maskings/Collection.cpp:46]:
```suggestion
    auto key = entry.key.stringView();
```

File: lib/Maskings/Collection.cpp [hunk 2]
```diff
@@ -23,23 +23,26 @@
...
-                                          VPackSlice const& def) {
+                                          velocypack::Slice def) {
   if (!def.isObject()) {
     return ParseResult<Collection>(
         ParseResult<Collection>::PARSE_FAILED,
         "expecting an object for collection definition");
   }
 
-  std::string type = "";
+  std::string type;
```
Comment (MBkkt) [inline lib/Maskings/Collection.cpp:42]: Also can be string_view

File: lib/Maskings/AttributeMasking.h [hunk 3]
```diff
@@ -34,34 +34,38 @@
...
-  bool match(std::vector<std::string> const&) const;
+  bool match(std::vector<std::string_view> const&) const;
 
   MaskingFunction* func() const { return _func.get(); }
 
  private:
   static std::unordered_map<std::string,
                             ParseResult<AttributeMasking> (*)(
-                                Path, Maskings*, VPackSlice const&)>
+                                Path, Maskings*, velocypack::Slice)>
```
Comment (MBkkt) [inline lib/Maskings/AttributeMasking.h:68]: Please make 
```
using Func =  ParseResult<AttributeMasking> (*)(
                                Path, Maskings*, velocypack::Slice);
```
You write this at least 3 time

File: lib/Maskings/AttributeMasking.cpp [hunk 4]
```diff
@@ -37,22 +37,22 @@ void arangodb::maskings::InstallMaskings() {
...
   if (!def.isObject()) {
     return ParseResult<AttributeMasking>(
         ParseResult<AttributeMasking>::PARSE_FAILED,
         "expecting an object for collection definition");
   }
 
-  std::string path = "";
-  std::string type = "";
+  std::string path;
+  std::string type;
```
Comment (MBkkt) [inline lib/Maskings/AttributeMasking.cpp:52]: Looks unused
```suggestion
```

File: lib/Maskings/AttributeMasking.cpp [hunk 5]
```diff
@@ -37,22 +37,22 @@ void arangodb::maskings::InstallMaskings() {
...
+                                                      velocypack::Slice def) {
   if (!def.isObject()) {
     return ParseResult<AttributeMasking>(
         ParseResult<AttributeMasking>::PARSE_FAILED,
         "expecting an object for collection definition");
   }
 
-  std::string path = "";
-  std::string type = "";
+  std::string path;
```
Comment (MBkkt) [inline lib/Maskings/AttributeMasking.cpp:51]: Also string_view?

File: lib/Maskings/Maskings.cpp [hunk 6]
```diff
@@ -200,112 +200,106 @@ bool Maskings::shouldDumpData(std::string const& name) {
...
-
-  for (auto const& entry : VPackObjectIterator(data, false)) {
-    std::string key = entry.key.copyString();
-    VPackSlice const& value = entry.value;
+                               std::vector<std::string_view>& path,
+                               velocypack::Slice data, VPackBuilder& out,
+                               std::string& buffer) const {
+  for (auto entry : VPackObjectIterator(data, false)) {
+    auto key = entry.key.stringView();
+    VPackSlice value = entry.value;
```
Comment (MBkkt) [inline lib/Maskings/Maskings.cpp:267]: NIT I don't sure but can we use same style
```suggestion
    velocypack::Slice value = entry.value;
```

**Review by MBkkt — APPROVED** [review]

File: lib/Maskings/AttributeMasking.cpp (hunk 4 above)
Comment (jsteemann) [reply lib/Maskings/AttributeMasking.cpp:52]: Both `_type` and `_path` are used later.

**Pull request merged**